euphoria-core/src/main/java/cz/seznam/euphoria/core/client/operator/Sort.java

```diff
@@ -0,0 +1,303 @@
+      return !partitioning.hasDefaultPartitioner() || partitioning.getNumPartitions() == 1;
+    }
+  }
+
+  public static <I> ByBuilder<I> of(Dataset<I> input) {
+    return new ByBuilder<>("Sort", input);
+  }
+
+  public static OfBuilder named(String name) {
+    return new OfBuilder(name);
+  }
+
+  // ~ -----------------------------------------------------------------------------
+
+  private final UnaryFunction<IN, S> sortByFn;
+
+  Sort(Flow flow,
+            String name,
+            Dataset<IN> input,
+            UnaryFunction<IN, S> sortByFn,
+            Partitioning<S> partitioning,
+            @Nullable Windowing<IN, W> windowing,
+            @Nullable ExtractEventTime<IN> eventTimeAssigner) {
+    super(name, flow, input, 
+        new PartitionKeyExtractor<>(sortByFn, partitioning), 
+        windowing, eventTimeAssigner, 
+        new HashPartitioning<>(partitioning.getNumPartitions()));
+    
+    this.sortByFn = sortByFn;
+  }
+
+  public UnaryFunction<IN, S> getSortByExtractor() {
+    return sortByFn;
+  }
+
+  @Override
+  public DAG<Operator<?, ?>> getBasicOps() {
+    Flow flow = getFlow();
+    
+    StateSupport.MergeFromStateMerger<IN, IN, Sorted<IN>>
+            stateCombiner = new StateSupport.MergeFromStateMerger<>();
+    ReduceStateByKey<IN, IN, IN, Integer, IN, Integer, IN, Sorted<IN>, W>
+        reduce =
+        new ReduceStateByKey<>(getName() + "::ReduceStateByKey", flow, input,
+                keyExtractor,
+                e -> e,
+                windowing,
+                eventTimeAssigner,
+                (StateFactory<IN, IN, Sorted<IN>>)
+                    (ctx, provider) -> new Sorted<>(ctx, provider, new SortByComparator<>(sortByFn)),
+                stateCombiner,
+                partitioning);
+
+    MapElements<Pair<Integer, IN>, IN>
+        format =
+        new MapElements<>(getName() + "::MapElements", flow, reduce.output(),
+            e -> e.getSecond());
+
+    DAG<Operator<?, ?>> dag = DAG.of(reduce);
+    dag.add(format, reduce);
+    return dag;
+  }
+  
+  private static class SortByComparator<V, S extends Comparable<S>> implements Comparator<V>, Serializable {
+
+    private final UnaryFunction<V, S> sortByFn;
+    
+    public SortByComparator(UnaryFunction<V, S> sortByFn) {
+      this.sortByFn = sortByFn;
+    }
+
+    @Override
+    public int compare(V o1, V o2) {
+      return sortByFn.apply(o1).compareTo(sortByFn.apply(o2));
+    }
+  }
+  
+  private static class PartitionKeyExtractor<IN, S extends Comparable<S>> implements UnaryFunction<IN, Integer> {
+
+    private final UnaryFunction<IN, S> sortByFn;
+    private final Partitioner<S> partitioner;
+    private final int numPartitions;
+    
+    public PartitionKeyExtractor(UnaryFunction<IN, S> sortByFn, Partitioning<S> partitioning) {
+      this.sortByFn = sortByFn;
+      this.partitioner = partitioning.getPartitioner();
+      this.numPartitions = partitioning.getNumPartitions();
```

Comment:
yes, this will not work. at the moment i'm a bit clueless how to go about it.

@horkyada may i ask you to add a corresponding comment to #92 so that we consider and fix it?